design/deploy-steps.md

```diff
@@ -0,0 +1,149 @@
+ This work is licensed under a Creative Commons Attribution 3.0
+ Unported License.
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+-->
+
+# Customizable deployment procedure
+
+## Status
+
+implementable
+
+## Summary
+
+This design proposes customizing the deployment procedure by providing a high
+level abstraction around Ironic [deploy steps][deploy-steps], as well as
+allowing per-host deployment (IPA) images.
+
+[deploy-steps]: https://docs.openstack.org/ironic/latest/admin/node-deployment.html#deploy-steps
+
+## Motivation
+
+By default the Ironic agent downloads a provided image via HTTP(s), converts it
+(if needed) and writes directly to the disk. This procedure works for many
+distributions, but may not suit some cases well. A prominent example is CoreOS,
+which has its own installer ([coreos-installer][coreos-installer]).
+
+[coreos-installer]: https://github.com/coreos/coreos-installer
+
+### Goals
+
+- A way to replace the default deploy procedure with a custom action.
+
+### Non-Goals
+
+- Expose the full functionality of Ironic deploy steps.
+- Add a way to execute custom actions during deployment.
+- A way to provide a link to the Ironic agent image per host.
+
+## Proposal
+
+### User Story
+
+As a CoreOS user I want to use `coreos-installer` to write the CoreOS image
+without losing any functionality that Ironic offers (unlike the live ISO
+workflow).
+
+## Design Details
+
+Additions to `BareMetalHost`:
+
+- New structure field `CustomDeploy` with the only field `Method` to engage
+  a non-default deployment procedure as described below.
+- The `Image.URL` field will become optional if `CustomDeploy` is set
+  since we don't know if the deploy method actually requires a separate image.
+
+Setting `CustomDeploy` to a non-empty value will trigger deployment even if
+`Image.URL` is empty.
+
+### Implementation Details: CustomDeploy
+
+When `CustomDeploy` is set, BMO will set the new [custom-agent deploy
+interface][custom-agent] on the node so that the default deploy steps do not
+apply. Then we request a deploy step with the name provided in the
+`CustomDeploy.Method` field. This step has to be provided in the agent image
+and is responsible for the whole deployment process.
+
+The `Image.URL` will no longer be required since some images (such as CoreOS)
+already include the root filesystem.
+
+NOTE: `CustomDeploy` is a single-field structure to allow potential extensions
+in the future. The `UserData` and `NetworkData` will be available to a deploy
+step implementation as part of the Node's `instance_info["configdrive"]`.
+Any required customization is expected to be provided via `UserData` and
+`NetworkData` or be built into the image itself.
+
+The expected workflow is as follows:
+
+1. A user sets `CustomDeploy.Method = "step_name"` and optionally `Image.URL`,
+   `UserData` and `NetworkData`.
+2. BMO sets `Node.DeployInterface = "custom-deploy"` and requests deployment
+   with a custom deploy step `step_name`.
```

Comment:
In the `Provisioning` state I guess

Reply:
Yeah, same place where the deployment logic (incl. live ISO) lives now.